### PR TITLE
Cosmos DB Spark Connector - Hierarchical Partitioning Key Support

### PR DESCRIPTION
**Pull request by TheovanKraay**
# Description

Adding support for hierarchical partition keys in Cosmos DB Spark Connector

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala [hunk 1]
```diff
@@ -192,6 +195,39 @@ private[cosmos] object SparkBridgeImplementationInternal extends BasicLoggingTra
     rangeToNormalizedRange(feedRange.getEffectiveRange(pkDefinition))
   }
 
+  private[cosmos] def partitionKeyValuesToNormalizedRange
+  (
+      partitionKeyValue: Object,
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala:200]: Why is this an object? Sounds like the expectation is that whaetverPassedHere toString retruns valid json - which is not happening in general. So, this API contract should be tightened - either by making the caller do the transformation to json and expect a json string here by narrowing down the types that can be used to ensure their toString representation is always json

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala [hunk 2]
```diff
@@ -192,6 +195,39 @@ private[cosmos] object SparkBridgeImplementationInternal extends BasicLoggingTra
...
+          for (value <- partitionKeyValues) {
+              partitionKey.add(value.trim)
+          }
+          partitionKey.build()
+          isMultiplePartitionKeyPaths = true
+      } catch {
+          case e: Exception =>
+              logInfo("Invalid partition key paths: " + json, e)
+      }
+      if (!isMultiplePartitionKeyPaths) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala:219]: Figuring out whether this is a single or multiple partition key by trying and catching exception is super inefficient - same as above here - please tighten the API contract

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala [hunk 3]
```diff
@@ -192,6 +195,39 @@ private[cosmos] object SparkBridgeImplementationInternal extends BasicLoggingTra
...
+          partitionKey.add(partitionKeyValue.toString.trim)
+      }
+
+      val feedRange = FeedRange
+          .forLogicalPartition(partitionKey.build())
+          .asInstanceOf[FeedRangePartitionKeyImpl]
+
+      val pkDefinition = SparkModelBridgeInternal.createPartitionKeyDefinitionFromJson(partitionKeyDefinitionJson)
+      rangeToNormalizedRange(feedRange.getEffectiveRange(pkDefinition))
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala:228]: NIT - LF before line 228 to make it easier to grasp the return Value

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala (hunk 2 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala:219]: Oh, I don't know why I have this check. I think this is from when I tried to make changes to the existing UDF, instead of creating a new one for multi-partition keys. Will remove.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala (hunk 1 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala:200]: Ok

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala [hunk 4]
```diff
@@ -169,17 +169,22 @@ private[spark] case class CosmosCatalogCosmosSDKClient(cosmosAsyncClient: Cosmos
 
     private def getPartitionKeyDefinition(containerProperties: Map[String, String]): PartitionKeyDefinition = {
         val partitionKeyPath = CosmosContainerProperties.getPartitionKeyPath(containerProperties)
-
         val partitionKeyDef = new PartitionKeyDefinition
         val paths = new util.ArrayList[String]
-        paths.add(partitionKeyPath)
+        val pathList = partitionKeyPath.split(",").toList
+        if (pathList.size > 1) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala:175]: Please gate the if-path on PKDefinition being >= 2 - the pkdefintion is available from CosmosConatinerProperties.getPartitionKeyVersion - the newly created PKDef should also have the same version as the original one - not hard-coded 2

File: sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml [hunk 5]
```diff
@@ -184,10 +184,5 @@
       <artifactId>jackson-module-scala_2.12</artifactId>
       <version>2.13.5</version> <!-- {x-version-update;com.fasterxml.jackson.module:jackson-module-scala_2.12;external_dependency} -->
     </dependency>
-    <dependency>
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-4_2-12/pom.xml:187]: Is this change necessary for Spark 3.1-3.3. as well or did we only add it to 3.4 accidentally?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import com.azure.cosmos.spark.CosmosPredicates.requireNotNullOrEmpty
+import org.apache.spark.sql.api.java.UDF2
+
+
+@SerialVersionUID(1L)
+class GetFeedRangeForPartitionKeyValues extends UDF2[String, Object, String] {
+  override def call
+  (
+    partitionKeyDefinitionJson: String,
+    partitionKeyValue: Object
```
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala]: changing this to array makes sense, but validating it is > 1... this is a new UDF and it is assumed it will only be used for multi partition key. Are you suggesting you prefer a single UDF, and suggesting this as a way of doing that cleanly? Won't it be a breaking change to make this array of objects in that case?

[the remark this answers is not in the document]
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala]: I looked at this, it seems Spark UDFs can take only strings as inputs, so it will have to expect json serialized string or an Object. Let me know if there is some other way (I can't find it).

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala (hunk 4 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala]: Ok, done.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala (hunk 3 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala]: Done

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala (hunk 2 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala]: Done

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala (hunk 1 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala]: I believe this has to be an object or string from the UDF.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala (hunk 4 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala]: Ok - this function is called to convert a generic key/value dictionary of string representing ConatinaerProperties into a "proper" PartitionKeyDefinition. There are two or three properties working together here PKPath (which you parse form the map) - but also version and kind.

With HPK I think your code looks good - except that you "blindly" use the eixstence of KeyPath with >1 entry as trigger to set version to 2 and kind to multi-hash. What if the incoming map has version==1 or kind==hash. In the absence of those properties in the map setting them to 2 and multihash is fine - but there needs to be a violation that would result in an error if the map has incosistent values. Does that make sense?

Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala]: makes sense, changes and corresponding tests added

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala]: LGTM now

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala]: https://stackoverflow.com/questions/59026439/how-to-create-a-spark-udf-in-java-which-accepts-array-of-strings

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala]: See this for an example - https://stackoverflow.com/questions/59026439/how-to-create-a-spark-udf-in-java-which-accepts-array-of-strings

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala]: I tried this, but when calling the UDF from Spark SQL, where the query text includes a call to a registered UDF, can only seem to pass in strings to the UDF parameters.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala [hunk 7]
```diff
@@ -169,17 +169,39 @@ private[spark] case class CosmosCatalogCosmosSDKClient(cosmosAsyncClient: Cosmos
...
+                            throw new IllegalArgumentException("PartitionKeyVersion V1 is not supported for multi-hash partition key")
+                        }
+                        PartitionKeyDefinitionVersion.V2
+                    }
+                case None => PartitionKeyDefinitionVersion.V2
+            })
+            pathList.foreach(path => paths.add(path.trim))
+        } else {
+            partitionKeyDef.setKind(PartitionKind.HASH)
+            paths.add(partitionKeyPath)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala:198]: maybe move line 178-181 logic to here? as for multi-hash, the version has already been configured.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/TestUtils.scala [hunk 8]
```diff
@@ -321,6 +322,8 @@ trait CosmosContainerWithRetention extends CosmosContainer {
   }
 }
 
+
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/TestUtils.scala:325]: nit: remove extra lines

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala [hunk 9]
```diff
@@ -184,12 +186,43 @@ private[cosmos] object SparkBridgeImplementationInternal extends BasicLoggingTra
     partitionKeyDefinitionJson: String
   ): NormalizedRange = {
 
+
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/implementation/SparkBridgeImplementationInternal.scala:189]: nit: remove extra line

**Comment by FabianMeiswinkel** [comment]
I also stumbled across few HPK comments in CosmosPatchTestHelper.scala - @TheovanKraay can you please check and if necessary reach out to @xinlian12 to get clarification if that is something else that needs to be covered?

**Comment by TheovanKraay** [comment]
> I also stumbled across few HPK comments in CosmosPatchTestHelper.scala - @TheovanKraay can you please check and if necessary reach out to @xinlian12 to get clarification if that is something else that needs to be covered?

I guess you meant "TODO: Reexamine the logic here when hierarchical partitioning being supported". I added a comment that this is not used/circumvented, see logic in getPatchItemWithSchema(). @xinlian12 can you let me know if you think this is ok?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/TestUtils.scala (hunk 8 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/TestUtils.scala]: thanks, done

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala (hunk 7 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/catalog/CosmosCatalogCosmosSDKClient.scala:210]: Ok, added the logic hash to there, and also moved set version so it is not done redundantly in the case of sub-partition.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+// Licensed under the MIT License.
+package com.azure.cosmos.spark.udf
+
+import com.azure.cosmos.implementation.SparkBridgeImplementationInternal
+import com.azure.cosmos.spark.CosmosPredicates.requireNotNullOrEmpty
+import org.apache.spark.sql.api.java.UDF2
+
+
+@SerialVersionUID(1L)
+class GetFeedRangeForPartitionKeyValues extends UDF2[String, String, String] {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala:11]: Currently, we have a UDF called` GetFeedRangeForPartitionKeyValue extends UDF2[String, Object, String]` 

I am wondering should we rename this udf to make it more obvious that it is for hierarchical partition keys or we combine these two udf into one, because I think the logic here also applies to the hash partition key
@FabianMeiswinkel  @TheovanKraay

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  override def call
+  (
+    partitionKeyDefinitionJson: String,
+    partitionKeyValue: String
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/GetFeedRangeForPartitionKeyValues.scala:15]: and also for the partitionKeyValue, should we constraint into only String type? because by theory, it can be String, Boolean, Number etc

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/PointWriterSubpartitionITest.scala [hunk 12]
```diff
@@ -0,0 +1,1382 @@
...
+
+    pointWriter.scheduleWrite(partitionKey, itemWithFullSchema)
+    pointWriter.flushAndClose()
+    // make sure the item exists
+    val originalItem = container.readItem(id, partitionKey, classOf[ObjectNode]).block().getItem()
+
+    // Cosmos patch does not support for system properties
+    // if we send request to patch for them, server is going to return exception with "Invalid patch request: Cannot patch system property"
+    // so the test is to make sure we have skipped these properties and the request can succeed for other properties
+    val partialUpdateSchema = StructType(Seq(
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/PointWriterSubpartitionITest.scala:704]: Can we also add a validation for the partitionKeyPath -> the path property should be ignored as well

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/utils/CosmosPatchTestHelper.scala [hunk 13]
```diff
@@ -219,7 +271,8 @@ object CosmosPatchTestHelper {
   *
   * @return the partition key path without
   */
-  // TODO: Reexamine the logic here when hierarchical partitioning being supported
+  //  TODO: Reexamine the logic here when hierarchical partitioning being supported
+  // for hierarchical partitioning this is not used/circumvented, see logic in getPatchItemWithSchema()
  def getStrippedPartitionKeyPath(partitionKeyDefinition: PartitionKeyDefinition): String = {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/utils/CosmosPatchTestHelper.scala:276]: can remove the TODO then

and also since the strippedPartitionKeyPath is not being used in the getPatchItemWithFullSchemaSubpartitions, I think it might be better to remove the partitionKeyPath parameter from the method

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/PointWriterSubpartitionITest.scala (hunk 12 above)
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/test/scala/com/azure/cosmos/spark/PointWriterSubpartitionITest.scala:700]: I've changed to ignore path for the subpartition tests.

**Review by TheovanKraay — COMMENTED** [review]
LGTM. Only nit I have is that I think the pk values in the tests should reflect what we have in Java SDK and Spark, i.e. "/tenantId", "/userId", "/sessionId", or at least a realistic hierarchy (id, firstName, LastName does not seem realistic). But I don't think this needs to block the PR.
